elementary/monitor/api/alerts/alerts.py

```diff
@@ -0,0 +1,277 @@
+            days_back, disable_samples
+        )
+        last_alert_sent_times = self._query_last_test_alert_times(days_back)
+        alerts_to_skip, alerts_to_send = self._sort_alerts(
+            pending_test_alerts, last_alert_sent_times
+        )
+        self._skip_alerts(
+            alerts_to_skip=alerts_to_skip, table_name=TestAlert.TABLE_NAME
+        )
+        return alerts_to_send
+
+    def get_model_alerts(self, days_back: int):
+        pending_model_alerts = self._query_pending_model_alerts(days_back)
+        last_alert_sent_times = self._query_last_model_alert_times(days_back)
+        alerts_to_skip, alerts_to_send = self._sort_alerts(
+            pending_model_alerts, last_alert_sent_times
+        )
+        self._skip_alerts(
+            alerts_to_skip=alerts_to_skip, table_name=ModelAlert.TABLE_NAME
+        )
+        return alerts_to_send
+
+    def get_source_freshness_alerts(self, days_back: int):
+        pending_source_freshness_alerts = self._query_pending_source_freshness_alerts(
+            days_back
+        )
+        last_alert_sent_times = self._query_last_source_freshness_alert_times(days_back)
+        alerts_to_skip, alerts_to_send = self._sort_alerts(
+            pending_source_freshness_alerts, last_alert_sent_times
+        )
+        self._skip_alerts(
+            alerts_to_skip=alerts_to_skip, table_name=SourceFreshnessAlert.TABLE_NAME
+        )
+        return alerts_to_send
+
+    def _sort_alerts(
+        self,
+        pending_alerts: AlertsQueryResult[Alert],
+        last_alert_sent_times: Dict[str, str],
+    ) -> Tuple[List[str], AlertsQueryResult[Alert]]:
+        suppressed_alerts = self._get_suppressed_alerts(
+            pending_alerts, last_alert_sent_times
+        )
+        alerts_to_skip = []
+        alerts_to_send = []
+        malformed_alerts_to_send = []
+
+        for alert in pending_alerts.alerts:
+            if alert.id in suppressed_alerts:
+                alerts_to_skip.append(alert.id)
+            else:
+                alerts_to_send.append(alert)
+
+        for alert in pending_alerts.malformed_alerts:
+            if alert.id in suppressed_alerts:
+                alerts_to_skip.append(alert.id)
+            else:
+                malformed_alerts_to_send.append(alert)
+
+        return alerts_to_skip, AlertsQueryResult(
+            alerts=alerts_to_send, malformed_alerts=malformed_alerts_to_send
+        )
+
+    def _get_suppressed_alerts(
+        self, alerts: AlertsQueryResult[Alert], last_alert_sent_times: Dict[str, str]
+    ) -> List[str]:
+        suppressed_alerts = []
+        current_time_utc = datetime.utcnow()
+        for alert in alerts.alerts:
+            id = (
+                alert.test_unique_id
+                if isinstance(alert, TestAlert)
+                else alert.unique_id
+            )
+            suppression_interval = alert.alert_suppression_interval
+            last_sent_time = (
```

Comment:
I don't exactly understand the concept of a last sent time.
This implies that an alert can be sent more than a single time.
Care to elaborate on what does that mean?

Reply:
So I guess that my naming is bad 😅 
Each test (unique on id column for model users for example) can create several alerts over the time (fail after fail after fail).
Each alert has its sent time, and we want o know for each test (model / source freshness / etc) what is the latest sent time.
We use that latest sent time to calculate the suppression 🙂